### DISCUSSION TITLE
v5, createPersister and IndexedDB: absolutely amazing!

### DISCUSSION THREAD

**Discussion by frederikhors** [TanStack/query#6213] [Show and tell] [answered]
I'm using the latest v5 version: AMAZING!!!!!!! Thanks!

I tried the `experimental_createPersister` and it works amazingly too!

I'm using it like this, what do you think?

```ts
import {
  experimental_createPersister,
  type AsyncStorage,
} from "@tanstack/query-persist-client-core";
import { get, set, del, createStore, type UseStore } from "idb-keyval";

function newIdbStorage(idbStore: UseStore): AsyncStorage {
  return {
    getItem: async (key) => await get(key, idbStore),
    setItem: async (key, value) => await set(key, value, idbStore),
    removeItem: async (key) => await del(key, idbStore),
  };
}

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      gcTime: 1000 * 30, // 30 seconds
      persister: !browser
        ? undefined
        : experimental_createPersister({
            storage: newIdbStorage(createStore("db_name", "store_name")),
            maxAge: 1000 * 60 * 60 * 12, // 12 hours
          }),
    },
  },
});
```

**Comment by TkDodo** [comment, marked as answer]
looks lovely

**Comment by TheTimeWalker** [comment]
Hey! Thank you so much for the snippet. Works pretty great!

One suggestion: The advantage of IndexedDB is that you can safely put objects into the storage without having to rely on serializing/deserializing the content like e.g. with LocalStorage. I have noticed that right now it would by default still stringify the content and then save it to IndexedDB. My suggestion would be to additionally add this into your persister:

```javascript
experimental_createPersister({
  storage: newIdbStorage(createStore("db_name", "store_name")),
  maxAge: 1000 * 60 * 60 * 12, // 12 hours,
  // Just pass the data
  serialize: (persistedQuery) => persistedQuery,
  deserialize: (cached) => cached,
})
```

The only problem is that TypeScript will complain that this serialize/deserialize function is not assignable, because it expects a string as a return type and cached type.

**Reply by frederikhors** [reply]
Amazing! I will try in a few minutes. Thanks!

**Reply by frederikhors** [reply]
It works amazingly well! Thanks!

@TheTimeWalker do you have any number about differences between JSON.string/parse and this? Have you had a chance to measure?

Yeah, Typescript complains about this. @TkDodo any suggestion? (I can open a PR for the docs)

**Reply by TkDodo** [reply]
What's the TS complaint? Can you show a typescript playground?

**Reply by TheTimeWalker** [reply]
@TkDodo Specifically this:
```typescript
Type '(persistedQuery: PersistedQuery) => PersistedQuery' is not assignable to type '(persistedQuery: PersistedQuery) => string'.
  Type 'PersistedQuery' is not assignable to type 'string'.ts(2322)
```
and
```typescript
Type '(cached: string) => string' is not assignable to type '(cachedString: string) => PersistedQuery'.
  Type 'string' is not assignable to type 'PersistedQuery'.ts(2322)
```
  
  Here's the [Typescript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUwB5mVUyB2MCGANgPoDGUy+MyACpgM7D3VRwC+cAZlBCHAOQABAjmb5SAawD0ARwCumAJ4BaDFEbNlpQsFwwt0ZPwDcAKFCRYiODEUY4AQXqKcpAMoxo+AObJ2XHj4hETFJWQUoFTUNfW1dPAMKE3NwaHgkXxgAGjh6ZGy4ABNkQhzbewBVPI9DHPJKahqKf25eAWBCgCNlCWRFADciZNNSCFF4eqpkAEkupp8-AF44AAoOzqbkAC44KuRNgEodpxd3TygFuEWAPkRTODgKGDkoHDuHh8zp6hAd-GdXKteooDldbvgAO74YDwTIrYE5daHLL3D55GDfZC-OD-U5Avo5QaEBSgm44qEw3L5eEEuBEhSIubnZAHFEfR5YiD9GY-P4A0j4kFg8nQ+DFQg0xSMjbM1mothmBWmEZjZhwaJMFhXVak66otBqbB4IhkChTOjqTWYFYIVEPZheXw7SbUWYyx3IFYu-bMlb8QpUfCdf5GHL8UYgEBjfgHOXsvJYIjAABe21WGuYyEKAEUIkKyRnqDm82yPsUE8Ak6mdl7xAALLO6uCketZ0tsA5mUwGtJFZCcfByQjwQuYMxAA)

@frederikhors 
I haven't done any measurements there. I'm also honestly not sure how to best measure this 🙈 I just assume that there could be a bit of a boost since you're not doing any transformations. Would be neat to know if that's the case at all. The biggest advantage is however that it makes navigating and looking at objects in DevTools nicer since you don't have a big escaped JSON string 😄

**Reply by TkDodo** [reply]
@TheTimeWalker the implementation in the playground seems wrong. `serialize` must return a `string`, but your version just returns what you pass in? That's not right ...

**Reply by TheTimeWalker** [reply]
Because the IndexedDB storage is used which means that it doesn't make much sense to return a string. If you serialize to string then the DB will only contain one entry which is just a stringified and escaped JSON object. After all, that's what's [said as an advantage in the persistQueryClient doc](https://tanstack.com/query/latest/docs/react/plugins/persistQueryClient#building-a-persister) where you _don't need_ to serialize the result :)

Unfortunately, the only way I see is to set the serializer and deserializer to just pass the object overwriting the `JSON.stringify` function as I can't just disable the serializer and natively pass the object to the database.

**Reply by TkDodo** [reply]
Oh, I didn't know that. So let's change the types to unknown instead of string? Or maybe a generic that defaults to string so that you can at least change it

**Reply by TheTimeWalker** [reply]
If I understand this correctly, if the string gets changed to unknown, then technically any type would be a possible return value which would remove type safety, right? I could guess that maybe that's not something you'd want. However, when I think about this, how necessary would it be to even have proper type safety there? With IndexedDB specifically, my serialization function would pretty much look like this:
```typescript
serialize: (persistedQuery): PersistedQuery  => persistedQuery
```
Since the query is really generic with any objects flowing through, I don't see how there could be any other definition than string or PersistedQuery. And if you end up with PersistedQuery in the end, it could as well be more beneficial to just somehow pass over and disable the serializer all together instead of having this pass through function

**Reply by TkDodo** [reply]
I'm thinking a generic that defaults to `string` should work so that at lest the input to `deserialize` is typed by whatever `serialize` returns.

**Reply by TkDodo** [reply]
one problem is that we default to `JSON.stringify` and `JSON.parse`, so if you pass one of the methods but not the other, typescript complains. If you can figure it out, I'll be happy to merge it :)

**Reply by TheTimeWalker** [reply]
I think that shouldn't be a problem 👍🏻
Since if you change the serializer to not do stringify, you'll have to change the deserializer anyway because it would try to parse a string by default. 😛

**Reply by TkDodo** [reply]
yes, but we need to get this right on type level. This shouldn't err at runtime

**Reply by frederikhors** [reply]
@TkDodo @TheTimeWalker What do you think if I open an RFC in the issues to discuss it better?

**Reply by TkDodo** [reply]
Yes please 👍

**Comment by TheTimeWalker** [comment]
With the newest version (v5.10), you can now add a type to the persister, so now we can combine my serializer solution with your IndexedDB solution :)

```typescript
import {
  experimental_createPersister,
  type AsyncStorage,
  type PersistedQuery
} from "@tanstack/query-persist-client-core";
import { get, set, del, createStore, type UseStore } from "idb-keyval";

function newIdbStorage(idbStore: UseStore): AsyncStorage<PersistedQuery> {
  return {
    getItem: async (key) => await get(key, idbStore),
    setItem: async (key, value) => await set(key, value, idbStore),
    removeItem: async (key) => await del(key, idbStore),
  };
}

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      gcTime: 1000 * 30, // 30 seconds
      persister: !browser
        ? undefined
        : experimental_createPersister<PersistedQuery>({
            storage: newIdbStorage(createStore("db_name", "store_name")),
            maxAge: 1000 * 60 * 60 * 12, // 12 hours,
            serialize: (persistedQuery) => persistedQuery,
            deserialize: (cached) => cached,
          }),
    },
  },
});
```

Under AsyncStorage and createrPersister, just add ``<PersistedQuery>``

**Reply by frederikhors** [reply]
Amazing! Amazing you, tanstack's team, all the open source world!

**Comment by tam-eoh** [comment]
When I use persistQueryClient, my data immediately appears upon reloading. However, when I switch to experimental_createPersister, my data is not there when reload. Could this mean I've set up something incorrectly?

![image](https://github.com/TanStack/query/assets/91381458/0740107f-a252-4a5c-acb4-374b7f8286ab)

**Reply by TkDodo** [reply]
That is the intended behavior because the data is only loaded when a useQuery call needs it. That is more memory efficient because not everything is loaded upfront. If you want to load everything upfront, the PersitQueryClient is what you want. Both approaches have their pros and cons.

**Reply by hverlin** [reply]
Re-reading this question, seems similar to my question: https://github.com/TanStack/query/discussions/6213#discussioncomment-8208175

Is the solution to use both `persister` together? Or use `setQueryData` for critical queries if we want to avoid any loading states?

**Comment by joelbrewer** [comment]
I must be doing something wrong, but I'm getting the following error when using `experimental_createPersister` :

<img width="1691" alt="image" src="https://github.com/TanStack/query/assets/2348386/422893f5-9a5a-4582-b1ff-90d699b77763">

Here's my code:

```
...
import { QueryClient } from '@tanstack/react-query';
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client'
import { experimental_createPersister } from "@tanstack/query-persist-client-core";
import { get, set, del, createStore } from "idb-keyval";
...

function newIdbStorage(idbStore) {
  return {
    getItem: async (key) => await get(key, idbStore),
    setItem: async (key, value) => await set(key, value, idbStore),
    removeItem: async (key) => await del(key, idbStore),
  };
}

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      gcTime: 1000 * 30, // 30 seconds
    },
  },
});

const Root = () => {
  const persister = experimental_createPersister({
    storage: newIdbStorage(createStore("my_db", "my_store")),
    maxAge: 1000 * 60 * 60 * 12, // 12 hours
  }); 
 
...

  return (
    <UserProvider>
      <PersistQueryClientProvider
        client={queryClient}
        persistOptions={{ persister }}
      >
        <RouterProvider router={router}>
          <ReactQueryDevtools initialIsOpen={false} />
        </RouterProvider>
      </PersistQueryClientProvider>
    </UserProvider>
  );
}

export default Root;
```

**Reply by TkDodo** [reply]
`experimental_createPersister` is not something you can pass to `PersistQueryClientProvider`

**Comment by hverlin** [comment]
@frederikhors @TkDodo @TheTimeWalker Thank you for the suggested solutions here!

I have an additional question regarding this new `experimental_createPersister` API, do you restore the cache immediately when you load the app?

```ts
import { createStore, del, get, set, UseStore, keys, getMany } from 'idb-keyval';

export async function restoreQueryClientCache() {
  if (!idbStore) {
    return;
  }

  const cachedKeys = await keys(idbStore);
  const cachedValues = await getMany<PersistedQuery>(cachedKeys, idbStore);
  for (const queryData of cachedValues) {
    if (queryData?.buster !== UI_VERSION) {
      continue;
    }

    if (!queryData?.state?.data) {
      continue;
    }

    queryClient.setQueryData(queryData.queryKey, queryData.state.data, {
      updatedAt: queryData.state.dataUpdatedAt,
    });
  }
}

restoreQueryClientCache().catch(console.error);

root.render(...);
```

I am migrating from the solution which was suggested [here](https://tanstack.com/query/latest/docs/react/plugins/persistQueryClient#building-a-persister).

<details>

```ts
import { get, set, del } from 'idb-keyval'
import { PersistedClient, Persister } from '@tanstack/react-query-persist-client'

export function createIDBPersister(idbValidKey: IDBValidKey = 'reactQuery') {
  return {
    persistClient: async (client: PersistedClient) => { await set(idbValidKey, client) },
    restoreClient: async () => { return await get<PersistedClient>(idbValidKey) },
    removeClient: async () => { await del(idbValidKey) },
  } as Persister
}
```

</details>

With the previous solution, the query would initially resolve with the data from the cache (which was quite nice as if one reloads the page, they immediately see the previous data - no loading state). After migrating to the new `experimental_createPersister`, the query will initially have no data until it reads from index db, leading to some flickering.

Is a workaround to run `queryClient.setQueryData` for all the queries stored in the cache before calling React render, as otherwise deserialization is lazy and will lead to flickering?

Wondering if there would be an option to make the query client immediately restore some of the queries?

**Reply by frederikhors** [reply]
@hverlin just to be clearer, are you talking about the loading that may occur between when the application starts and when the data is fetched from the cache?

**Reply by hverlin** [reply]
yes. Since fetching the data from the cache is async, if you don't use `setQueryData` you get some very short loading state

**Comment by AnandChowdhary** [comment]
Here is my solution, in case it helps someone. It doesn't write to IndexedDB immediately after each request, instead doing it when there's idle time and no more than once every 10 seconds. It also doesn't wait more than a minute and guarantees writing to cache before unload and when changing tabs, so I feel like it's the best balance.

```ts
import {
  experimental_createPersister,
  type AsyncStorage,
  type PersistedQuery,
} from "@tanstack/query-persist-client-core";
import { QueryClient } from "@tanstack/react-query";
import { createStore, del, get, set, type UseStore } from "idb-keyval";

/**
 * Creates an IndexedDB storage for React Query persistence
 *
 * This storage:
 * - Groups writes together to improve performance
 * - Uses browser idle time when available
 * - Handles different queue sizes with different strategies
 * - Ensures data is saved before page unload or tab switch
 *
 * @param idbStore - The IndexedDB store from idb-keyval
 * @returns Storage compatible with TanStack Query persister
 */
function newIdbStorage(idbStore: UseStore): AsyncStorage<PersistedQuery> {
  let pendingWrites = new Map<string, PersistedQuery>();
  let isProcessing = false;
  let idleCallbackId: number | null = null;
  let maxDelayTimer: number | null = null;
  let lastProcessTime = 0;

  const MAX_DELAY = 60000; // 1 minute max before forced processing
  const CRITICAL_THRESHOLD = 100; // Number of writes that triggers immediate processing
  const THROTTLE_DELAY = 10000; // 10 seconds between processing

  /**
   * Process all pending writes to IndexedDB
   *
   * Steps:
   * 1. Gather all pending writes
   * 2. Clear scheduled timers
   * 3. Write items to IndexedDB in parallel
   * 4. Re-queue any failed writes
   */
  const processPendingWrites = async () => {
    if (pendingWrites.size === 0 || isProcessing) return;

    const now = Date.now();
    if (now - lastProcessTime < THROTTLE_DELAY) {
      // Too soon to process again, reschedule
      scheduleWrites();
      return;
    }

    console.debug("Processing pending writes");

    try {
      isProcessing = true;
      lastProcessTime = now;
      const writes = Array.from(pendingWrites.entries());
      pendingWrites.clear();

      if (idleCallbackId) {
        if ("requestIdleCallback" in window) window.cancelIdleCallback(idleCallbackId);
        idleCallbackId = null;
      }

      if (maxDelayTimer) {
        clearTimeout(maxDelayTimer);
        maxDelayTimer = null;
      }

      await Promise.all(
        writes.map(async ([key, value]) => {
          try {
            await set(key, value, idbStore);
          } catch (error) {
            console.error(`Failed to persist query for key ${key}:`, error);
            // Re-add failed writes for retry
            pendingWrites.set(key, value);
          }
        }),
      );
    } finally {
      isProcessing = false;
      if (pendingWrites.size > 0) scheduleWrites();
    }
  };

  /**
   * Schedule processing of pending writes
   *
   * Uses two ways:
   * 1. requestIdleCallback for when browser is not busy
   * 2. setTimeout as a backup to ensure it happens anyway
   */
  const scheduleWrites = () => {
    // Skip if already scheduled
    if (idleCallbackId || maxDelayTimer) return;

    const timeUntilNextProcess = Math.max(0, THROTTLE_DELAY - (Date.now() - lastProcessTime));

    if ("requestIdleCallback" in window) {
      idleCallbackId = window.requestIdleCallback(() => processPendingWrites(), {
        timeout: MAX_DELAY, // Ensure callback runs within this time even if browser is busy
      }) as unknown as number;
    }

    // Ensure we don't process more often than every 10 seconds
    maxDelayTimer = window.setTimeout(processPendingWrites, Math.max(timeUntilNextProcess, MAX_DELAY));
  };

  // Set up event listeners for persistence guarantees
  if (typeof window !== "undefined") {
    // Process writes before page unload
    window.addEventListener(
      "beforeunload",
      (event) => {
        if (pendingWrites.size > 0) {
          // May trigger browser confirmation in some browsers
          event.preventDefault();
          // Force processing regardless of throttle on page unload
          lastProcessTime = 0;
          processPendingWrites();
        }
      },
      { capture: true }, // Run before other handlers
    );

    // Process writes when tab becomes hidden
    // Ensures data is saved when switching tabs or minimizing
    window.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "hidden" && pendingWrites.size > 0) {
        // Force processing regardless of throttle when tab becomes hidden
        lastProcessTime = 0;
        processPendingWrites();
      }
    });
  }

  return {
    getItem: async (key) => {
      if (pendingWrites.has(key)) return pendingWrites.get(key);
      return await get(key, idbStore);
    },

    /**
     * Set an item in storage
     *
     * Handles storage differently based on queue size:
     * 1. Small queue: Wait for idle time or up to 60s
     * 2. Large queue (≥100 items): Save immediately
     */
    setItem: async (key, value) => {
      pendingWrites.set(key, value);

      if (pendingWrites.size >= CRITICAL_THRESHOLD) {
        // Process immediately when queue gets too large
        // May briefly slow UI but prevents memory problems
        const now = Date.now();
        if (now - lastProcessTime >= THROTTLE_DELAY) {
          await processPendingWrites();
        } else {
          scheduleWrites();
        }
      } else scheduleWrites();
    },

    removeItem: async (key) => {
      pendingWrites.delete(key);
      await del(key, idbStore);
    },
  };
}

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      gcTime: 1000 * 30, // After 30 seconds read from persistence
      persister:
        typeof window === "undefined"
          ? undefined
          : experimental_createPersister<PersistedQuery>({
              storage: newIdbStorage(createStore("db_name", "store_name")),
              maxAge: Infinity, // Keep persisted queries forever
              serialize: (persistedQuery) => persistedQuery,
              deserialize: (cached) => cached,
            }),
    },
  },
});

```

**Reply by frederikhors** [reply]
I think this is what `experimental_createPersister` does already, nope?

**Comment by spookyuser** [comment]
I think it might be a good idea to include this in experimental_createQueryPersister

```ts
function newIdbStorage(idbStore: UseStore): AsyncStorage {
  return {
    getItem: async (key) => await get(key, idbStore),
    setItem: async (key, value) => await set(key, value, idbStore),
    removeItem: async (key) => await del(key, idbStore),
  };
}
```

It feels like it's basically recommended everywhere experimental_createQueryPersister is and if you define it slightly differently to this, typescript goes crazy

**Comment by luca992** [comment]
Fyi, for others, using `idb-keyval` isn't the best,  `onversionchange` isn't handled at all by `idb-keyval`. If you have more than one tab open and delete the database in one tab, the delete will wait for all other tabs to close that have opened an IDB connection with `idb-keyval` before continuing. Making all react query request freeze until those tabs close.

My current solution:

```typescript
import {
  type AsyncStorage,
  experimental_createQueryPersister,
  type PersistedQuery,
} from "@tanstack/query-persist-client-core";

const DB_NAME = "db-name";
const STORE_NAME = "store";
const VERSION = 1;

let dbPromise: Promise<IDBDatabase> | null = null;

function getDb(): Promise<IDBDatabase> {
  if (dbPromise) return dbPromise;

  dbPromise = new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, VERSION);

    req.onupgradeneeded = () => {
      const upgradeDb = req.result;
      if (!upgradeDb.objectStoreNames.contains(STORE_NAME)) {
        upgradeDb.createObjectStore(STORE_NAME);
      }
    };

    req.onsuccess = () => {
      const conn = req.result;
      conn.onversionchange = () => {
        console.trace("[IDB] version change – closing connection");
        conn.close();
        // clear the cached promise so the next getDb call re-opens a connection
        dbPromise = null;
      };
      resolve(conn);
    };

    req.onblocked = () => {
      console.warn("[IDB] open blocked: another connection is holding the DB");
    };

    req.onerror = () => {
      const err = req.error;
      reject(
        err instanceof Error
          ? err
          : new Error(`IndexedDB open failed: ${String(err)}`),
      );
      dbPromise = null;
    };
  });

  return dbPromise;
}

const storage: AsyncStorage<PersistedQuery> = {
  async getItem(key: string) {
    const db = await getDb();
    return new Promise<PersistedQuery | undefined>((res, rej) => {
      const r = db
        .transaction(STORE_NAME, "readonly")
        .objectStore(STORE_NAME)
        .get(key);
      r.onsuccess = () => res(r.result as PersistedQuery);
      r.onerror = () => rej(r.error ?? new Error("getItem failed"));
    });
  },

  async setItem(key: string, value: PersistedQuery) {
    return getDb()
      .then((db) => {
        if (!db) return;
        return new Promise<void>((resolve, reject) => {
          const req = db
            .transaction(STORE_NAME, "readwrite")
            .objectStore(STORE_NAME)
            .put(value, key);
          req.onsuccess = () => resolve();
          req.onerror = () => reject(req.error ?? new Error("setItem failed"));
        });
      })
      .catch((err) => {
        // experimental_createQueryPersister expects no error to be thrown
        console.trace("[IDB] skipping setItem:", err);
      });
  },
  
  async removeItem(key: string) {
    return getDb()
      .then((db) => {
        if (!db) return;
        return new Promise<void>((resolve, reject) => {
          const req = db
            .transaction(STORE_NAME, "readwrite")
            .objectStore(STORE_NAME)
            .delete(key);
          req.onsuccess = () => resolve();
          req.onerror = () =>
            reject(req.error ?? new Error("removeItem failed"));
        });
      })
      .catch((err) => {
        // experimental_createQueryPersister expects no error to be thrown
        console.trace(
          "[IDB] skipping removeItem:",
          err,
        );
      });
  },
};

export const persister = experimental_createQueryPersister<PersistedQuery>({
  storage,
  maxAge: 1000 * 60 * 60 * 24 * 30, // 30 days
  serialize: (persistedQuery) => persistedQuery,
  deserialize: (cached) => cached,
});
```

**Reply by frederikhors** [reply]
Your code is amazing, @luca992.

I created the [entries method](https://tanstack.com/query/v5/docs/framework/react/plugins/createPersister) too:

```ts
async entries(): Promise<[string, PersistedQuery][]> {
    try {
        const db = await getDb();
        const transaction = db.transaction(STORE_NAME, 'readonly');
        const store = transaction.objectStore(STORE_NAME);
        const keysRequest = store.getAllKeys();
        const valuesRequest = store.getAll();

        const [keys, values] = await Promise.all([
            promisifyRequest(keysRequest),
            promisifyRequest(valuesRequest)
        ]);

        return keys.map((key, index) => [String(key), values[index]]);
    } catch (error) {
        console.error('failed entries:', error);
        return [];
    }
}
```

with:

```ts
function promisifyRequest<T>(request: IDBRequest<T>): Promise<T> {
	return new Promise((resolve, reject) => {
		request.onsuccess = () => resolve(request.result);
		request.onerror = () => reject(request.error ?? new Error('IndexedDB operation failed'));
	});
}
```

What do you think?

**Reply by luca992** [reply]
Looks solid to me 👍.

**Reply by luca992** [reply]
@frederikhors One thing, I think entries is safe to throw errors from (similar to how functions using `getItem` all handle errors), since the functions using entries already throws another error.

https://github.com/TanStack/query/blob/a1b1279be3befc37a9d4fbb2fa79826bfb0955a4/packages/query-persist-client-core/src/createPersister.ts#L224-L284


Unlike set and remove which silently broke `experimental_createQueryPersister` if they threw errors in my tests

**Reply by frederikhors** [reply]
@luca992 so how do you think it's best to change it?

**Reply by luca992** [reply]
Just saying its probably safe to throw an error there if you want instead of an empty array on error

**Reply by frederikhors** [reply]
Do you mean this?

```ts
async entries(): Promise<[string, PersistedQuery][]> {
    try {
        const db = await getDb();
        const transaction = db.transaction(STORE_NAME, 'readonly');
        const store = transaction.objectStore(STORE_NAME);
        const keysRequest = store.getAllKeys();
        const valuesRequest = store.getAll();

        const [keys, values] = await Promise.all([
            promisifyRequest(keysRequest),
            promisifyRequest(valuesRequest)
        ]);

        return keys.map((key, index) => [String(key), values[index]]);
    } catch (error) {
        console.error('failed entries:', error);

        throw error;
    }
}
```

**Reply by luca992** [reply]
Just 

```typescript
async entries(): Promise<[string, PersistedQuery][]> {
        const db = await getDb();
        const transaction = db.transaction(STORE_NAME, 'readonly');
        const store = transaction.objectStore(STORE_NAME);
        const keysRequest = store.getAllKeys();
        const valuesRequest = store.getAll();

        const [keys, values] = await Promise.all([
            promisifyRequest(keysRequest),
            promisifyRequest(valuesRequest)
        ]);

        return keys.map((key, index) => [String(key), values[index]]);
}
```

**Reply by luca992** [reply]
But also, It'd be safer to do with a single cursor, something like:

```typescript
async entries(): Promise<[string, PersistedQuery][]> {
  const db = await getDb();
  const tx = db.transaction(STORE_NAME, 'readonly');
  const store = tx.objectStore(STORE_NAME);

  const out: [string, PersistedQuery][] = [];

  await new Promise<void>((resolve, reject) => {
    tx.oncomplete = () => resolve();
    tx.onerror = () => reject(tx.error);

    const req = store.openCursor();
    req.onerror = () => reject(req.error);
    req.onsuccess = () => {
      const cursor = req.result as IDBCursorWithValue | null;
      if (!cursor) return; // done
      out.push([String(cursor.key), cursor.value]);
      cursor.continue();
    };
  });

  return out;
}
```

**Reply by frederikhors** [reply]
Works amazingly well! Thanks!

**Comment by jakearchibald** [comment]
> Fyi, for others, using idb-keyval isn't the best, onversionchange isn't handled at all by idb-keyval.

`idb-keyval` only ever opens a single version of the database, so it never has a version change.

> If you have more than one tab open and delete the database in one tab

Yeah, `idb-keyval` assumes it is in control of the database, and it doesn't offer the ability to delete the database. If you want lower-level control, use https://www.npmjs.com/package/idb.

**Reply by frederikhors** [reply]
@jakearchibald what do you tyink about @luca992's [code](https://github.com/TanStack/query/discussions/6213#discussioncomment-13623303)?

**Reply by luca992** [reply]
Just proposing to use `idb-keyval` here is risky, and can lead to people running into bugs, like that locking bug I mentioned, that are really really hard to debug. It abstracts away the db too much. And doesn't handle what I think is a common scenario where you just want to wipe all data for a user when they click logout in a web-app (what I ran into). Also.. I'm not adding a dependency to idb to my project just to wrap a callback in a promise, if I can just you know... interact with idb itself in like 10 lines of code no dependencies needed

**Reply by jakearchibald** [reply]
`idb-keyval` does offer a `clear()` method. https://github.com/jakearchibald/idb-keyval?tab=readme-ov-file#clear

I kinda feel @luca992 has made up their mind here, regardless of evidence.

https://www.npmjs.com/package/idb exists for lower-level control. I guess it's true that it just "wraps a callback in a promise", but in the same way that "React just wraps DOM mutations". Also it's 1.2kb.

**Reply by frederikhors** [reply]
Thanks @jakearchibald.

> `idb-keyval` does offer a `clear()` method.

What happens if more tabs are open and one of them calls `clear()`? Because this is the issue we are trying to solve here. Right now everything freezes and we must close all the tabs opened.

Is there a way to reload all the tabs if idb-keyval's `clear()` is called?

**Reply by jakearchibald** [reply]
> What happens if more tabs are open and one of them calls `clear()`? Because this is the issue we are trying to solve here. Right now everything freezes and we must close all the tabs opened.

I really don't think that's the case. Is there a minimal reproduction of this issue?

**Reply by frederikhors** [reply]
@luca992 do you know how to arrange a minimal reproduction of the issue?

**Reply by luca992** [reply]
@jakearchibald "made up my mind, regardless of evidence", idk what high horse you think you're on but this is not a discussion on your library. If I wanted to discuss issues with your library I would have submitted an issue in your repo about the fact that `idb-keyval` does not handle recovering from `onversionchange` when the db it uses is deleted, which is a bug in your implementation that could 100% be handled.

Plus, 1.2kb is not nothing, and I had no plans to use it anywhere else in my project. There's no need for a dependency to make a `experimental_createPersister` instance backed by idb.

I'm sure `idb-keyval` has it's uses but my point is it's not a great choice for this discussion which is creating a `experimental_createPersister` because of this edge case. I spent days figuring out why all react-query operations were stalling, and figured out it was due to `idb-keyval` holding open the db after deleting the dbs in another tab. I decided to share because I thought it would prevent other from having to deal with the pain of debugging this edge case. 

@frederikhors It's quite easy to reproduce. You just delete all indexDb databases in one tab, while `idb-keyval` holds it open in another. All further `idb-keyval` requests will stall until the tab holding the database open with `idb-keyval` is closed

**Reply by jakearchibald** [reply]
https://random-stuff.jakearchibald.com/idb-keyval/clear/ - here's a basic example. Data can be set & cleared with a single command, even if there are multiple tabs.

**Reply by luca992** [reply]
I am well aware clear would work, that's not what I said

**Reply by jakearchibald** [reply]
I'm replying to https://github.com/TanStack/query/discussions/6213#discussioncomment-14221291

**Comment by anhdd-kuro** [comment]
@jakearchibald @frederikhors @luca992 
I'm so confused right now. Can you guys provide the full setup code with and without idb-keyval using latest version, please?  
[The official docs only mention this code](https://tanstack.com/query/v5/docs/framework/react/plugins/persistQueryClient#building-a-persister)  

```ts
import { get, set, del } from 'idb-keyval'
import {
  PersistedClient,
  Persister,
} from '@tanstack/react-query-persist-client'

/**
 * Creates an Indexed DB persister
 * @see https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API
 */
export function createIDBPersister(idbValidKey: IDBValidKey = 'reactQuery') {
  return {
    persistClient: async (client: PersistedClient) => {
      await set(idbValidKey, client)
    },
    restoreClient: async () => {
      return await get<PersistedClient>(idbValidKey)
    },
    removeClient: async () => {
      await del(idbValidKey)
    },
  } satisfies Persister
}
```

Is this enough? Why do you need to use experimental_createQueryPersister then?  
And as I can see, `experimental_createQueryPersister` is still experimental after 2 years, which is not a good sign to use it

**Reply by TkDodo** [reply]
the docs mention what you can do to create your own persister, and shows index db with idb-keyval as an example. You can _also_ use our `AsyncPersister` and provide `idb-keyval` as a storage. You probably don’t need to build your own persister just for index db.

> with and without idb-keyval

no idea how index-db works without a wrapper. horrible api 🤷‍♂️.

> Why do you need to use experimental_createQueryPersister

Let’s back up a bit. `persistQueryClient` or `<PersistQueryClient>` needs a `persister`, and they always persist the complete `QueryClient`.

`createQueryPersister` is a different way of persisting - it wraps the `queryFn` of a single Query and lazily reads that data (from a single storage per query) when the query is first used. It’s a totally different thing and you don’t combine them - you choose either or. It came out of this discussion:

- https://github.com/TanStack/query/discussions/2649
